tempodb/sharding_block_selector.go

```diff
@@ -0,0 +1,264 @@
+		MaxCompactionRange:   maxCompactionRange,
+		MaxCompactionObjects: maxCompactionObjects,
+		MaxBlockBytes:        maxBlockBytes,
+		boundaries:           traceidboundary.All(uint32(shards)),
+	}
+
+	var (
+		now        = time.Now()
+		currWindow = s.windowForTime(now)
+	)
+
+	for _, b := range blocklist {
+
+		var (
+			w     = s.windowForBlock(b)
+			shard = s.shardOf(b.MinID)
+
+			// These are all of the numeric values that we can group and order by,
+			shardMin = fmt.Sprintf("%03d", shard)
+			shardMax = fmt.Sprintf("%03d", s.shardOf(b.MaxID))
+			columns  = fmt.Sprintf("%016X", b.DedicatedColumnsHash())
+			level    = fmt.Sprintf("%03d", b.CompactionLevel)
+			window   = fmt.Sprintf("%v", w)
+			age      = fmt.Sprintf("%016X", currWindow-w)
+			min      = fmt.Sprintf("%X", b.MinID)
+		)
+
+		entry := shardingBlockEntry{
+			meta:      b,
+			minBlocks: s.MinInputBlocks,
+			// Within group order by lowest compaction level first,
+			// then lowest trace ID to try to co-locate traces even further.
+			// This is the same for all cases.
+			order: join(level, min),
+		}
+
+		switch {
+		case b.CompactionLevel > 0 && shardMin == shardMax:
+			// This is a previously split block that contains only a single shard.
+			// Combine/dedupe with other blocks in the same window and shard.
+			// This is prioritized over splitting new blocks to keep the block count down.
+			entry.group = join(
+				"A",            // Highest priority
+				age,            // prioritize newer windows (more recent data)
+				shardMin,       // same shard
+				b.Version,      //
+				b.DataEncoding, // same block format and column config
+				columns,        //
+			)
+			entry.hash = join(
+				b.TenantID, // Work sharding by tenant, window
+				window,     //
+				shardMin,   // and shard
+			)
+
+		case b.CompactionLevel == 0:
+			// Unsharded new block that needs to be split. This step generates work
+			// for the above step. MinBlocks=1 allows stray single blocks to still be split.
+			entry.split = true
+			entry.minBlocks = 1
+			entry.group = join(
+				"B",            // Second priority
+				age,            //
+				b.Version,      //
+				b.DataEncoding, //
+				columns,        //
+			)
+			entry.hash = join(
+				b.TenantID, // Work sharding by tenant and window
+				window,     //
+			)
+
+		default:
+			// Existing block without sharding, or sharded under different settings.
+			// Retroactively upgrade them but only if there is spare compactor
+			// capacity, by assigning them the lowest priority.
+			entry.split = true
+			entry.minBlocks = 1
+			entry.group = join(
+				"C",            // lowest priority
+				age,            //
+				b.Version,      //
+				b.DataEncoding, //
+				columns,        //
+			)
+			entry.hash = join(
+				b.TenantID, // Work sharding by tenant and window
+				window,     //
+			)
+		}
+
+		s.entries = append(s.entries, entry)
+	}
+
+	// sort by group then order
+	sort.SliceStable(s.entries, func(i, j int) bool {
```

Comment:
is stable important?

Reply:
This was inherited from the original block selector.  Don't believe it is, and not seeing any clues from the original [commit](https://github.com/grafana/tempo/commit/9264916552304513c4586b2828e3e736fc5fa384#diff-094e7acc4b5e77da6729bc4f464aac009e471a8e78a3c3db82eda52cb1c1be35R133).